### ISSUE TITLE
Crash on izCachedValueAndErrorForKey:createBlock

### ISSUE THREAD

**Issue by TomCarey** [iZettle/sdk-ios#126]
This might be related to low memory? But I have repeat occurrences of the following crash:

```
# Date: 2017-02-17T13:26:00Z
# OS Version: 10.2.1 (14D27)
# Device: iPad Mini 2
# RAM Free: 5.3%
# Disk Free: 86%

#0. Crashed: com.apple.main-thread
0   iZettleShared                  0x101bfe76c -[NSObject(CachedValue) izCachedValueAndErrorForKey:createBlock:] + 376
1   iZettlePayments                0x101a55eac -[DatecsReader consumeEvent] + 104
2   iZettlePayments                0x101a54ae4 -[DatecsReaderQueue initialPayloadForOperation:] + 212
3   iZettlePayments                0x101a35c78 -[CardReaderQueue(Operations) sendBackendOperation:withPayload:] + 72
4   iZettlePayments                0x101a32b50 __81-[CardReaderQueue(Operations) performBackendOperation:basePayload:resultHandler:]_block_invoke_3 + 160
5   iZettleShared                  0x101c20c80 __32-[IZFutureQueue executeNextItem]_block_invoke + 72
6   iZettleShared                  0x101c20bd8 -[IZFutureQueue executeNextItem] + 260
7   iZettleShared                  0x101c216e8 -[IZFutureQueue executeFunction:] + 656
8   iZettlePayments                0x101a32a5c __81-[CardReaderQueue(Operations) performBackendOperation:basePayload:resultHandler:]_block_invoke_2 + 188
9   iZettleShared                  0x101c21d58 __29-[IZFutureQueue executeBatch]_block_invoke_2 + 184
10  iZettleShared                  0x101c20c80 __32-[IZFutureQueue executeNextItem]_block_invoke + 72
11  iZettleShared                  0x101c20bd8 -[IZFutureQueue executeNextItem] + 260
12  iZettleShared                  0x101c216e8 -[IZFutureQueue executeFunction:] + 656
13  iZettleShared                  0x101c21b60 __29-[IZFutureQueue executeBatch]_block_invoke + 200
14  iZettlePayments                0x101a32960 __81-[CardReaderQueue(Operations) performBackendOperation:basePayload:resultHandler:]_block_invoke + 152
15  iZettleShared                  0x101c4a524 __28-[IZFuture(Operations) then]_block_invoke_2 + 64
16  iZettleShared                  0x101c49c94 __37-[IZFuture(Operations) thenComplete:]_block_invoke_2 + 84
17  iZettleShared                  0x101c49bd0 __37-[IZFuture(Operations) thenComplete:]_block_invoke.145 + 336
18  iZettleShared                  0x101c489d4 -[IZTask setCompletion:] + 288
19  iZettleShared                  0x101c497bc -[IZFuture(Operations) thenComplete:] + 552
20  iZettleShared                  0x101c4a488 __28-[IZFuture(Operations) then]_block_invoke + 132
21  iZettlePayments                0x101a32838 -[CardReaderQueue(Operations) performBackendOperation:basePayload:resultHandler:] + 416
22  iZettlePayments                0x101a40acc __161-[UINavigationController(PaymentFlow) processPaymentUsingAccount:amount:cardInterface:collectedData:paymentReferencesDictionary:accountType:numberOfInstalments:]_block_invoke.132 + 144
23  iZettleShared                  0x101c4a524 __28-[IZFuture(Operations) then]_block_invoke_2 + 64
24  iZettleShared                  0x101c49c94 __37-[IZFuture(Operations) thenComplete:]_block_invoke_2 + 84
25  iZettleShared                  0x101c49bd0 __37-[IZFuture(Operations) thenComplete:]_block_invoke.145 + 336
26  iZettleShared                  0x101c489d4 -[IZTask setCompletion:] + 288
27  iZettleShared                  0x101c497bc -[IZFuture(Operations) thenComplete:] + 552
28  iZettleShared                  0x101c4a488 __28-[IZFuture(Operations) then]_block_invoke + 132
29  iZettlePayments                0x101a40614 __161-[UINavigationController(PaymentFlow) processPaymentUsingAccount:amount:cardInterface:collectedData:paymentReferencesDictionary:accountType:numberOfInstalments:]_block_invoke.107 + 328
30  iZettleShared                  0x101c4a8f0 __35-[IZFuture(Operations) thenOrCatch]_block_invoke_2 + 56
31  iZettleShared                  0x101c49c94 __37-[IZFuture(Operations) thenComplete:]_block_invoke_2 + 84
32  iZettleShared                  0x101c49bd0 __37-[IZFuture(Operations) thenComplete:]_block_invoke.145 + 336
33  iZettleShared                  0x101c489d4 -[IZTask setCompletion:] + 288
34  iZettleShared                  0x101c497bc -[IZFuture(Operations) thenComplete:] + 552
35  iZettleShared                  0x101c4a85c __35-[IZFuture(Operations) thenOrCatch]_block_invoke + 132
36  iZettlePayments                0x101a400d4 __161-[UINavigationController(PaymentFlow) processPaymentUsingAccount:amount:cardInterface:collectedData:paymentReferencesDictionary:accountType:numberOfInstalments:]_block_invoke.29 + 1072
37  iZettlePayments                0x101a3f404 -[UINavigationController(PaymentFlow) processPaymentUsingAccount:amount:cardInterface:collectedData:paymentReferencesDictionary:accountType:numberOfInstalments:] + 772
38  iZettlePayments                0x101a724e0 __120-[UINavigationController(PurchaseFlow) chargeAccount:amount:showSettings:paymentReferencesFunction:postPaymentFunction:]_block_invoke.144 + 144
39  iZettleShared                  0x101c4a524 __28-[IZFuture(Operations) then]_block_invoke_2 + 64
40  iZettleShared                  0x101c49c94 __37-[IZFuture(Operations) thenComplete:]_block_invoke_2 + 84
41  iZettleShared                  0x101c49bd0 __37-[IZFuture(Operations) thenComplete:]_block_invoke.145 + 336
42  iZettleShared                  0x101c489d4 -[IZTask setCompletion:] + 288
43  iZettleShared                  0x101c497bc -[IZFuture(Operations) thenComplete:] + 552
44  iZettleShared                  0x101c4a488 __28-[IZFuture(Operations) then]_block_invoke + 132
45  iZettlePayments                0x101a721f8 __120-[UINavigationController(PurchaseFlow) chargeAccount:amount:showSettings:paymentReferencesFunction:postPaymentFunction:]_block_invoke.131 + 576
46  iZettleShared                  0x101c4a524 __28-[IZFuture(Operations) then]_block_invoke_2 + 64
47  iZettleShared                  0x101c49c94 __37-[IZFuture(Operations) thenComplete:]_block_invoke_2 + 84
48  iZettleShared                  0x101c49bd0 __37-[IZFuture(Operations) thenComplete:]_block_invoke.145 + 336
49  iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
50  iZettleShared                  0x101c4c150 __28+[IZFuture(Operations) all:]_block_invoke.249 + 660
51  iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
52  iZettleShared                  0x101c49eb0 __37-[IZFuture(Operations) thenComplete:]_block_invoke_2 + 624
53  iZettleShared                  0x101c49bd0 __37-[IZFuture(Operations) thenComplete:]_block_invoke.145 + 336
54  iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
55  iZettleShared                  0x101c4a014 __37-[IZFuture(Operations) thenComplete:]_block_invoke.150 + 108
56  iZettleShared                  0x101c489d4 -[IZTask setCompletion:] + 288
57  iZettleShared                  0x101c49ddc __37-[IZFuture(Operations) thenComplete:]_block_invoke_2 + 412
58  iZettleShared                  0x101c49bd0 __37-[IZFuture(Operations) thenComplete:]_block_invoke.145 + 336
59  iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
60  iZettleShared                  0x101c4ae80 __31-[IZFuture(Operations) always:]_block_invoke.209 + 120
61  iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
62  iZettleShared                  0x101c4cf64 __28+[IZFuture(Operations) any:]_block_invoke.267 + 428
63  iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
64  iZettleShared                  0x101c4a014 __37-[IZFuture(Operations) thenComplete:]_block_invoke.150 + 108
65  iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
66  iZettleShared                  0x101c4a014 __37-[IZFuture(Operations) thenComplete:]_block_invoke.150 + 108
67  iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
68  iZettleShared                  0x101c49eb0 __37-[IZFuture(Operations) thenComplete:]_block_invoke_2 + 624
69  iZettleShared                  0x101c49bd0 __37-[IZFuture(Operations) thenComplete:]_block_invoke.145 + 336
70  iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
71  iZettleShared                  0x101c4a014 __37-[IZFuture(Operations) thenComplete:]_block_invoke.150 + 108
72  iZettleShared                  0x101c489d4 -[IZTask setCompletion:] + 288
73  iZettleShared                  0x101c49ddc __37-[IZFuture(Operations) thenComplete:]_block_invoke_2 + 412
74  iZettleShared                  0x101c49bd0 __37-[IZFuture(Operations) thenComplete:]_block_invoke.145 + 336
75  iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
76  iZettleShared                  0x101c49eb0 __37-[IZFuture(Operations) thenComplete:]_block_invoke_2 + 624
77  iZettleShared                  0x101c49bd0 __37-[IZFuture(Operations) thenComplete:]_block_invoke.145 + 336
78  iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
79  iZettleShared                  0x101c6d294 -[IZTask(Utilities) complete:orAbort:onQueue:] + 160
80  iZettleShared                  0x101c210dc __32-[IZFutureQueue executeNextItem]_block_invoke.65 + 180
81  iZettleShared                  0x101c6da6c __38-[IZFuture(Utilities) onResultOrError]_block_invoke_2 + 56
82  iZettleShared                  0x101c4a8f0 __35-[IZFuture(Operations) thenOrCatch]_block_invoke_2 + 56
83  iZettleShared                  0x101c49c94 __37-[IZFuture(Operations) thenComplete:]_block_invoke_2 + 84
84  iZettleShared                  0x101c49bd0 __37-[IZFuture(Operations) thenComplete:]_block_invoke.145 + 336
85  iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
86  iZettleShared                  0x101c4ae80 __31-[IZFuture(Operations) always:]_block_invoke.209 + 120
87  iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
88  iZettleShared                  0x101c4a014 __37-[IZFuture(Operations) thenComplete:]_block_invoke.150 + 108
89  iZettleShared                  0x101c489d4 -[IZTask setCompletion:] + 288
90  iZettleShared                  0x101c49ddc __37-[IZFuture(Operations) thenComplete:]_block_invoke_2 + 412
91  iZettleShared                  0x101c49bd0 __37-[IZFuture(Operations) thenComplete:]_block_invoke.145 + 336
92  iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
93  iZettleShared                  0x101c4a014 __37-[IZFuture(Operations) thenComplete:]_block_invoke.150 + 108
94  iZettleShared                  0x101c489d4 -[IZTask setCompletion:] + 288
95  iZettleShared                  0x101c49ddc __37-[IZFuture(Operations) thenComplete:]_block_invoke_2 + 412
96  iZettleShared                  0x101c49bd0 __37-[IZFuture(Operations) thenComplete:]_block_invoke.145 + 336
97  iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
98  iZettleShared                  0x101c4a014 __37-[IZFuture(Operations) thenComplete:]_block_invoke.150 + 108
99  iZettleShared                  0x101c489d4 -[IZTask setCompletion:] + 288
100 iZettleShared                  0x101c49ddc __37-[IZFuture(Operations) thenComplete:]_block_invoke_2 + 412
101 iZettleShared                  0x101c49bd0 __37-[IZFuture(Operations) thenComplete:]_block_invoke.145 + 336
102 iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
103 iZettleShared                  0x101c49eb0 __37-[IZFuture(Operations) thenComplete:]_block_invoke_2 + 624
104 iZettleShared                  0x101c49bd0 __37-[IZFuture(Operations) thenComplete:]_block_invoke.145 + 336
105 iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
106 iZettleShared                  0x101c4ae80 __31-[IZFuture(Operations) always:]_block_invoke.209 + 120
107 iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
108 iZettleShared                  0x101c6d294 -[IZTask(Utilities) complete:orAbort:onQueue:] + 160
109 iZettleShared                  0x101c6d468 -[IZTask(Utilities) complete:orAbort:] + 64
110 iZettleShared                  0x101c70598 __59+[IZFuture(Utilities) performOnlyOneAtTheTimeForKey:block:]_block_invoke + 264
111 iZettleShared                  0x101c4a8f0 __35-[IZFuture(Operations) thenOrCatch]_block_invoke_2 + 56
112 iZettleShared                  0x101c49c94 __37-[IZFuture(Operations) thenComplete:]_block_invoke_2 + 84
113 iZettleShared                  0x101c49bd0 __37-[IZFuture(Operations) thenComplete:]_block_invoke.145 + 336
114 iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
115 iZettleShared                  0x101c4a014 __37-[IZFuture(Operations) thenComplete:]_block_invoke.150 + 108
116 iZettleShared                  0x101c489d4 -[IZTask setCompletion:] + 288
117 iZettleShared                  0x101c49ddc __37-[IZFuture(Operations) thenComplete:]_block_invoke_2 + 412
118 iZettleShared                  0x101c49bd0 __37-[IZFuture(Operations) thenComplete:]_block_invoke.145 + 336
119 iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
120 iZettleShared                  0x101c49eb0 __37-[IZFuture(Operations) thenComplete:]_block_invoke_2 + 624
121 iZettleShared                  0x101c49bd0 __37-[IZFuture(Operations) thenComplete:]_block_invoke.145 + 336
122 iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
123 iZettleShared                  0x101c49eb0 __37-[IZFuture(Operations) thenComplete:]_block_invoke_2 + 624
124 iZettleShared                  0x101c49bd0 __37-[IZFuture(Operations) thenComplete:]_block_invoke.145 + 336
125 iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
126 iZettleShared                  0x101c6d294 -[IZTask(Utilities) complete:orAbort:onQueue:] + 160
127 iZettleShared                  0x101c210dc __32-[IZFutureQueue executeNextItem]_block_invoke.65 + 180
128 iZettleShared                  0x101c6da6c __38-[IZFuture(Utilities) onResultOrError]_block_invoke_2 + 56
129 iZettleShared                  0x101c4a8f0 __35-[IZFuture(Operations) thenOrCatch]_block_invoke_2 + 56
130 iZettleShared                  0x101c49c94 __37-[IZFuture(Operations) thenComplete:]_block_invoke_2 + 84
131 iZettleShared                  0x101c49bd0 __37-[IZFuture(Operations) thenComplete:]_block_invoke.145 + 336
132 iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
133 iZettleShared                  0x101c4ae80 __31-[IZFuture(Operations) always:]_block_invoke.209 + 120
134 iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
135 iZettleShared                  0x101c4a014 __37-[IZFuture(Operations) thenComplete:]_block_invoke.150 + 108
136 iZettleShared                  0x101c489d4 -[IZTask setCompletion:] + 288
137 iZettleShared                  0x101c49ddc __37-[IZFuture(Operations) thenComplete:]_block_invoke_2 + 412
138 iZettleShared                  0x101c49bd0 __37-[IZFuture(Operations) thenComplete:]_block_invoke.145 + 336
139 iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
140 iZettleShared                  0x101c49eb0 __37-[IZFuture(Operations) thenComplete:]_block_invoke_2 + 624
141 iZettleShared                  0x101c49bd0 __37-[IZFuture(Operations) thenComplete:]_block_invoke.145 + 336
142 iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
143 iZettleShared                  0x101c49eb0 __37-[IZFuture(Operations) thenComplete:]_block_invoke_2 + 624
144 iZettleShared                  0x101c49bd0 __37-[IZFuture(Operations) thenComplete:]_block_invoke.145 + 336
145 iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
146 iZettleShared                  0x101c4cf64 __28+[IZFuture(Operations) any:]_block_invoke.267 + 428
147 iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
148 iZettleShared                  0x101c4ae80 __31-[IZFuture(Operations) always:]_block_invoke.209 + 120
149 iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
150 iZettleShared                  0x101c4a014 __37-[IZFuture(Operations) thenComplete:]_block_invoke.150 + 108
151 iZettleShared                  0x101c492f0 -[IZTask complete:orAbort:onScheduler:] + 384
152 iZettleShared                  0x101c6d294 -[IZTask(Utilities) complete:orAbort:onQueue:] + 160
153 iZettlePayments                0x101a567ac -[DatecsReader _receiveData:] + 608
154 iZettlePayments                0x101a2a184 -[AccessoryService stream:handleEvent:] + 560
155 ExternalAccessory              0x1960edff8 -[EAInputStream _streamEventTrigger] + 504
156 CoreFoundation                 0x1888eab5c __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 24
157 CoreFoundation                 0x1888ea4a4 __CFRunLoopDoSources0 + 524
158 CoreFoundation                 0x1888e80a4 __CFRunLoopRun + 804
159 CoreFoundation                 0x1888162b8 CFRunLoopRunSpecific + 444
160 GraphicsServices               0x18a2ca198 GSEventRunModal + 180
161 UIKit                          0x18e8567fc -[UIApplication _run] + 684
162 UIKit                          0x18e851534 UIApplicationMain + 208
163 MYapp                   0x100302500 main (main.m:65)
164 libdispatch.dylib              0x1877f95b8 (Missing)
```

From what I can tell this occurs in 1.4 and 1.4.1 versions of the SDK.

**Comment by TomCarey** [comment]
Sorry, just noticed there is a 1.4.2 version now, so I'll try that and update the issue if I see it occur again.

**Comment by mansbernhardt** [comment]
@TomCarey I don't think 1.4.2 solves this. This is a crash we are seeing in our app as well. @mattiasjahnke what is the current status of that issue?

**Comment by jtrotsky** [comment]
Hi all, any further info on the state of this issue? 
@mansbernhardt @mattiasjahnke

**Comment by mansbernhardt** [comment]
@jtrotsky We are still working on a fix. We have had a hard time to reproduce it. But we think we finally found a way. It seems to happen when you turn on the reader (or connect it) with a card already inserted, while the SDK is showing the connect reader screen. 
We have a potential fix, but need to do some regression testing as well as test it in the field with our own app first. So I can't give you a date when this will be released in an SDK update.

**Comment by jtrotsky** [comment]
Awesome, appreciate the good work @mansbernhardt

**Comment by jtrotsky** [comment]
Still seeing crashes for this, although infrequently. @mansbernhardt assume no further progress yet?

**Comment by mansbernhardt** [comment]
@jtrotsky Yes this bug has been fixed in our main app, but has not gone into a public SDK release yet. @mattiasjahnke correct me if I'm wrong.

**Comment by mattiasjahnke** [comment]
Sounds about right! We'll try to deliver a new SDK version asap

**Comment by jtrotsky** [comment]
Awesome @mattiasjahnke @mansbernhardt thanks for the quick update!

**Comment by jtrotsky** [comment]
Is this issue addressed in the v1.5 stability improvements? 🤞  @mattiasjahnke @mansbernhardt

**Comment by mattiasjahnke** [comment]
Yes that crash shown now be gone!
